tolstoy/src/tx_client.rs

```diff
@@ -0,0 +1,113 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use std::collections::BTreeMap;
+use std::collections::btree_map::Entry;
+
+use rusqlite;
+
+use errors::{
+    Result,
+    ErrorKind
+};
+
+use mentat_db::types::{
+    Entid
+};
+
+use mentat_db::{
+    entids,
+    TypedSQLValue
+};
+
+use mentat_core::{
+    TypedValue
+};
+
+#[derive(Debug)]
+pub struct TxPart {
+    pub e: Entid,
+    pub a: i64,
+    pub v: TypedValue,
+    pub added: i32
+}
+
+#[derive(Debug)]
+pub struct Tx {
+    pub tx: Entid,
+    pub tx_instant: TypedValue,
+    pub parts: Vec<TxPart>
```

Comment:
This is fine to start with, fine for experimentation, and it's probably fine for streaming in changes from a server, if you're not buffering them to disk.

If you _are_ buffering them to disk, or you're thinking ahead to a representation of a just-committed transaction, there are two other representations you should consider:

- A structure that makes typical tx-listener questions — "does this transaction mention an attribute or entity I care about?" — efficient. That might be a trie, it might be a bunch of extra data structures (e.g., a set of mentioned attributes), or something else. With an unsorted `Vec<TxPart>`, looking for a mentioned attribute requires linear search of the entire vector.

- A structure that doesn't require pulling the entire tx into memory. This might be a cursor, a rowid range, or something else that's scoped to the lifetime of a particular database transaction, in order to preserve isolation.